hoomd/mpcd/RandomSubsetPicker.h

```diff
@@ -0,0 +1,292 @@
+     * number of particles picked to N_pick indexes that were picked in the *original* flags array.
+     */
+
+    /*
+     * First calculating How many marked particles are there in flags(1 = marked) and storing their indexes in m_flags_idx
+     * This is done by CalculateTotalNumbersToPickFrom()
+     */
+    unsigned int N_total = CalculateTotalNumbersToPickFrom<T>(flags, N);
+    unsigned int N_total_all_ranks = N_total;
+    unsigned int N_before = 0;
+
+    // reduce / scan the number of particles that are marked on all ranks
+    #ifdef ENABLE_MPI
+    if (m_exec_conf->getNRanks() > 1)
+        {
+        MPI_Allreduce(MPI_IN_PLACE, &N_total_all_ranks, 1, MPI_UNSIGNED, MPI_SUM, m_exec_conf->getMPICommunicator());
+        MPI_Exscan(&N_total, &N_before, 1, MPI_UNSIGNED, MPI_SUM, m_exec_conf->getMPICommunicator());
+        }
+    #endif // ENABLE_MPI
+
+    // If N_try_pick >= N_total_all_ranks - pick all particles
+    if (N_total_all_ranks <= N_try_pick)
+        {
+        m_picks_idx.resize(N_total);
+        ArrayHandle<unsigned int> h_picks_idx(m_picks_idx, access_location::host, access_mode::overwrite);
+        std::iota(h_picks_idx.data, h_picks_idx.data + N_total, 0);
+        m_Npick = N_total;
+        }
+    else
+        {
+        // do the pick logic
+        makeAllPicks(timestep, N_try_pick, N_total_all_ranks);
+        /*
+         * Select the picks that lie on my rank, with reindexing to local mark indexes.
+         * This is performed in a do loop to allow for resizing of the GPUVector.
+         * After a short time, the loop will be ignored.
+         */
+        const unsigned int max_pick_idx = N_before + N_total;
+        bool overflowed = false;
+        do
+            {
+            m_Npick = 0;
+            const unsigned int max_Npick = m_picks_idx.getNumElements();
+
+                {
+                ArrayHandle<unsigned int> h_picks_idx(m_picks_idx, access_location::host, access_mode::overwrite);
+                for (unsigned int i=0; i < N_try_pick; ++i)
+                    {
+                    const unsigned int pick = m_all_picks[i];
+                    if (pick >= N_before && pick < max_pick_idx)
+                        {
+                        if (m_Npick < max_Npick)
+                            {
+                            h_picks_idx.data[m_Npick] = pick - N_before;
+                            }
+                        ++m_Npick;
+                        }
+                    }
+                }
+
+            overflowed = (m_Npick > max_Npick);
+            if (overflowed)
+                {
+                m_picks_idx.resize(m_Npick);
+                }
+
+            } while (overflowed);
+        }
+    //storing indexes of picked particles according to the *original* flags array in picks and updating N_pick
+    storePicksIdx(picks);
+    N_pick = m_Npick;
```

Comment:
I would have this `storePicksIdx` also assign the value to `N_pick` (rather than do separate), since they are closely related to each other.

You may want a better name for the function too, like `assignPicks`